File: Programmers/연습문제/Level 3/부대복귀/Solution.swift
import Foundation

func solution(_ n:Int, _ roads:[[Int]], _ sources:[Int], _ destination:Int) -> [Int] {

    var adjMatrix: [Int: [(Int, Int)]] = [:]
    var dist: [Int: Int] = [:]
    roads.forEach { road in
        if let start = road.first,
           let end = road.last {
              adjMatrix[start, default: []].append((end, 1))
              adjMatrix[end, default: []].append((start, 1))
        }
    }

    var queue: [(Int, Int)] = [(destination, 0)]
    var index = 0
    var visited = Array(repeating: false, count: n+1)
    visited[destination] = true

    while index < queue.count {
        let (currNode, currDist) = (queue[index].0, queue[index].1)
        index += 1
        dist[currNode] = currDist

        guard let postNodes = adjMatrix[currNode]
        else {
            dist[currNode] = -1
            continue
        }

        postNodes.forEach { (postNode, postDist) in
            if !visited[postNode] {
                visited[postNode] = true
                queue.append((postNode, currDist + postDist))
            }
        }
    }

    return sources.map { dist[$0] ?? -1 }
}

let n = 3
let roads = [[1, 2], [2, 3]]
let sources = [2, 3]
let destination = 1
print(solution(n, roads, sources, destination)) // [1, 2]

File: Programmers/연습문제/Level 3/부대복귀/Solution.py
import heapq


def solution(n, roads, sources, destination):
    adj = {i: [] for i in range(1, n+1)}
    for s, e, in roads:
        adj[s].append([e, 1])
        adj[e].append([s, 1])

    INF = float('inf')
    dist = [INF] * (n+1)
    dist[destination] = 0

    heap_queue = []
    heapq.heappush(heap_queue, [0, destination])
    while heap_queue:
        node = heapq.heappop(heap_queue)
        for end in adj[node[1]]:
            if dist[end[0]] > node[0] + end[1]:
                dist[end[0]] = node[0] + end[1]
                heapq.heappush(heap_queue, [dist[end[0]], end[0]])

    answer = []
    for source in sources:
        if dist[source] == INF:
            answer.append(-1)
        else:
            answer.append(dist[source])
    return answer


if __name__ == "__main__":
    n = 3
    roads = [[1, 2], [2, 3]]
    sources = [2, 3]
    destination = 1
    print(solution(n, roads, sources, destination))
